File: legacy/functionapp/models/terminal.py
"""Pydantic models shared across HTTP triggers and orchestrators."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class ProvisionTerminalRequest(BaseModel):
    """User-overridable inputs when provisioning a Remote Terminal VM."""

    subscription_id: str = Field(..., description="Subscription that will host the VM.")
    resource_group: str = Field("rg-elb-terminal", min_length=1, max_length=90)
    region: str = Field("koreacentral", min_length=2, max_length=40)
    vm_name: str = Field("vm-elb-terminal", min_length=1, max_length=15)
    vm_size: str = Field("Standard_D4s_v5")
    admin_username: str = Field("azureuser", min_length=3, max_length=32)
    allowed_ssh_cidr: str = Field(
        ..., description="Caller egress IP in CIDR form, e.g. 1.2.3.4/32."
    )
    # Optional — for auto-assigning RBAC roles to VM managed identity
    workload_resource_group: str = Field("", description="Workload RG for Contributor role.")
    acr_resource_group: str = Field("", description="ACR RG for AcrPull role.")
    acr_name: str = Field("", description="ACR name for AcrPull role.")
    storage_account: str = Field("", description="Storage account for Blob Data Contributor role.")
    storage_resource_group: str = Field("", description="Storage account RG.")


class ProvisionTerminalStarted(BaseModel):
    instance_id: str
    status_query_uri: str


class TerminalConnectionInfo(BaseModel):
    vm_name: str
    resource_group: str
    region: str
    fqdn: str
    ssh_host: str
    ssh_port: int = 22
    username: str
    password_secret_uri: str
    cloud_init_status: Literal["running", "done", "failed", "unknown"] = "unknown"


class HealthResponse(BaseModel):
    status: Literal["ok"] = "ok"
    version: str = "0.1.0"

File: web/src/api/terminal.ts
import { api } from "@/api/client";
import type { OrchestrationStatus } from "@/api/shared";

export interface ProvisionTerminalRequest {
  subscription_id: string;
  resource_group?: string;
  region?: string;
  vm_name?: string;
  vm_size?: string;
  admin_username?: string;
  allowed_ssh_cidr: string;
  workload_resource_group?: string;
  acr_resource_group?: string;
  acr_name?: string;
  storage_account?: string;
  storage_resource_group?: string;
}

export interface ProvisionTerminalStarted {
  id: string;
  statusQueryGetUri: string;
  sendEventPostUri: string;
  terminatePostUri: string;
}

export interface TerminalConnectionInfo {
  vm_name: string;
  resource_group: string;
  subscription_id?: string;
  region: string;
  fqdn: string;
  ssh_host: string;
  ssh_port: number;
  username: string;
  password_secret_uri: string;
  cloud_init_status: string;
}

export interface TerminalHealth {
  az_cli: string;
  kubectl: string;
  azcopy: string;
  python: string;
  az_login_active: boolean;
  az_login_user: string;
  az_login_age_seconds: number;
}

export const terminalApi = {
  provision: (req: ProvisionTerminalRequest) =>
    api.post<ProvisionTerminalStarted>("/terminal/provision", req),

  status: (instanceId: string) =>
    api.get<OrchestrationStatus<TerminalConnectionInfo>>(
      `/terminal/status/${encodeURIComponent(instanceId)}`,
    ),

  password: (vmName: string, subscriptionId?: string, resourceGroup?: string) => {
    const params = new URLSearchParams();
    if (subscriptionId) params.set("subscription_id", subscriptionId);
    if (resourceGroup) params.set("resource_group", resourceGroup);
    const qs = params.toString();
    return api.get<{ vm_name: string; password: string }>(
      `/terminal/${encodeURIComponent(vmName)}/password${qs ? `?${qs}` : ""}`,
    );
  },

  openSsh: (
    vmName: string,
    callerIp: string,
    subscriptionId: string,
    resourceGroup: string,
  ) =>
    api.post<{ ok: boolean; nsg: string; allowed_ip: string }>(
      `/terminal/${encodeURIComponent(vmName)}/open-ssh?caller_ip=${encodeURIComponent(callerIp)}&subscription_id=${encodeURIComponent(subscriptionId)}&resource_group=${encodeURIComponent(resourceGroup)}`,
      {},
    ),

  stopVm: (vmName: string, subscriptionId: string, resourceGroup: string) =>
    api.post<{ ok: boolean; vm_name: string; status: string }>(
      `/terminal/${encodeURIComponent(vmName)}/stop?subscription_id=${encodeURIComponent(subscriptionId)}&resource_group=${encodeURIComponent(resourceGroup)}`,
      {},
    ),

  startVm: (vmName: string, subscriptionId: string, resourceGroup: string) =>
    api.post<{ ok: boolean; vm_name: string; status: string }>(
      `/terminal/${encodeURIComponent(vmName)}/start?subscription_id=${encodeURIComponent(subscriptionId)}&resource_group=${encodeURIComponent(resourceGroup)}`,
      {},
    ),

  health: (vmName: string, subscriptionId: string, resourceGroup: string) =>
    api.get<TerminalHealth>(
      `/terminal/${encodeURIComponent(vmName)}/health?subscription_id=${encodeURIComponent(subscriptionId)}&resource_group=${encodeURIComponent(resourceGroup)}`,
    ),
};